### PR TITLE
codeintel: Cache tag resolution in GraphQL resolvers

### PR DESCRIPTION
**Pull request by efritz**
Code intel resolvers tend to return duplicate objects within a given repository (lists of code locations will often contain multiple copies of the same repository, commit, or path resolver). We have a tree of memoized "cached" repositories, commits, and path resolvers.

image

We neglected to cache the `Tags` resolution from commits, and if tags are requested from a commit object, it will re-request tags from gitserver regardless of it's the same commit resolver object or not in the backend (our underlying graphql library does not automatically memoize results, it seems).

This behavior is so far not new. But, in 5.0, we introduced a new resolver and UI pages in the SPA that are able to cause an explosion of `git tag` calls to a single gitserver shard. This may cause such a a large explosion of duplicate calls that it cripples gitserver from normal usage.

This PR adds memoization to the tags resolver along this code path so that each unique commit will only resolve tags once within a single graphql request (from precise index resolvers, at least).

## Test plan

Validated by hand that memoization works as intended. Uploaded > 1000 (fake) indexes to a small set of commits. Ensured that the tag resolution of each commit only happened once per page query.

image

## Relation to customer issues

This checks a lot of boxes we've seen related to customer issues recently:

- Affects monorepos with code intelligence uploaded
- Is involved with behavior introduced in 5.0 (the code graph dashboard pages)
- Is the only place where we're relevantly make a `git tags ... --points-at=<commit>` call

Where I'm not 100% convinced this is the root cause yet, 

but we're waiting on data from both instances to confirm this suspicion.

### CODE REVIEW

File: enterprise/internal/codeintel/uploads/transport/graphql/precise_index_resolver.go [hunk 1]
```diff
@@ -326,31 +314,27 @@ func (r *preciseIndexResolver) ProjectRoot(ctx context.Context) (_ resolverstubs
...
-	var (
-		repoName api.RepoName
-		commit   string
-	)
-	if r.upload != nil {
-		repoName, commit = api.RepoName(r.upload.RepositoryName), r.upload.Commit
-	} else if r.index != nil {
-		repoName, commit = api.RepoName(r.index.RepositoryName), r.index.Commit
-	}
+	// defer r.traceErrs.Collect(&err, log.String("uploadResolver.field", "tags"))
```
Comment (Numbers88s) [inline enterprise/internal/codeintel/uploads/transport/graphql/precise_index_resolver.go:317]: should we just remove this?

Comment (efritz) [reply enterprise/internal/codeintel/uploads/transport/graphql/precise_index_resolver.go:317]: It's a good reminder that maybe we can put it back someday :laff: You got it, will get rid of it pre-merge.

**Review by Numbers88s — APPROVED** [review]

**Review by eseliger — APPROVED** [review]

**Review by kopancek — COMMENTED** [review]
Can we add some unit tests testing the new behavior?

**Comment by efritz** [comment]
Instructions to populate an instance representative of affected monorepo instances are provided in [this gist](https://gist.github.com/efritz/4fa3321b112e88130b91979c8f6a8f6c). Use any repository and any (small) SCIP index as only the metadata matters here. Navigate to the dashboard page of your chosen repo.

Screenshot 2023-06-12 at 1 36 56 PM

We see the same symptoms as large monorepos instances, where a flood of git tag requests peg CPU for a long period (even after the dashboard page is closed). Running on this branch gives a much more reasonable set of tag calls.

Screenshot 2023-06-12 at 1 45 17 PM

**Pull request merged**